### ISSUE TITLE
Scanning order with recurse

### ISSUE THREAD

**Issue by XhmikosR** [JustinBeckwith/linkinator#45]
It seems the order of links is quite different when recurse is used.

It's like when a link is discovered in crawl mode, it jumps to that page immediately and then comes back to the original file that referenced the link.

Wouldn't it make more sense to finish each page and then for each link that is discovered, follow it?

Just thinking out loud here, trying to see why I get so different results with and without recurse :)

Example of the current behavior:

```
> bootstrap@4.3.1 linkinator C:\Users\xmr\Desktop\bootstrap
> linkinator _gh_pages/ --skip "^(?!http://localhost)"

🏊‍♂️ crawling _gh_pages/
  [200] http://localhost:5773

 Scanning http://localhost:5773
  [SKP] https://getbootstrap.com/
  [404] http://localhost:5773/docs/4.3/dist/css/bootstrap.css
  [200] http://localhost:5773/docs/4.3/assets/css/docs.min.css
  [200] http://localhost:5773/docs/4.3/assets/img/favicons/apple-touch-icon.png
  [200] http://localhost:5773/docs/4.3/assets/img/favicons/favicon-32x32.png
  [200] http://localhost:5773/docs/4.3/assets/img/favicons/favicon-16x16.png
  [200] http://localhost:5773/docs/4.3/assets/img/favicons/manifest.json
  [200] http://localhost:5773/docs/4.3/assets/img/favicons/safari-pinned-tab.svg
  [200] http://localhost:5773/favicon.ico
  [200] http://localhost:5773/
  [200] http://localhost:5773/docs/4.3/getting-started/introduction/
  [200] http://localhost:5773/docs/4.3/examples/
  [SKP] https://themes.getbootstrap.com/
  [SKP] https://expo.getbootstrap.com/
  [SKP] https://blog.getbootstrap.com/
  [200] http://localhost:5773/docs/4.3/
  [SKP] https://getbootstrap.com/docs/4.2/
  [SKP] https://getbootstrap.com/docs/4.0/
  [SKP] https://v4-alpha.getbootstrap.com/
  [SKP] https://getbootstrap.com/docs/3.4/
  [SKP] https://getbootstrap.com/docs/3.3/
  [SKP] https://getbootstrap.com/2.3.2/
  [200] http://localhost:5773/docs/versions/
  [SKP] https://github.com/twbs/bootstrap
  [SKP] https://twitter.com/getbootstrap
  [SKP] https://bootstrap-slack.herokuapp.com/
  [SKP] https://opencollective.com/bootstrap/
  [200] http://localhost:5773/docs/4.3/getting-started/download/
  [SKP] https://www.bootstrapcdn.com/
  [200] http://localhost:5773/docs/4.3/layout/overview/
  [200] http://localhost:5773/docs/4.3/about/overview/
  [200] http://localhost:5773/docs/4.3/about/team/
  [SKP] https://github.com/twbs/bootstrap/graphs/contributors
  [SKP] https://github.com/twbs/bootstrap/blob/master/LICENSE
  [SKP] https://creativecommons.org/licenses/by/3.0/
  [SKP] https://www.google-analytics.com/analytics.js
  [SKP] https://cdn.carbonads.com/carbon.js?serve=CKYIKKJL&placement=getbootstrapcom
  [200] http://localhost:5773/docs/4.3/assets/img/bootstrap-themes.png
  [SKP] https://code.jquery.com/jquery-3.3.1.slim.min.js
  [404] http://localhost:5773/docs/4.3/dist/js/bootstrap.bundle.js
  [200] http://localhost:5773/docs/4.3/assets/js/vendor/anchor.min.js
  [200] http://localhost:5773/docs/4.3/assets/js/vendor/clipboard.min.js
  [200] http://localhost:5773/docs/4.3/assets/js/vendor/bs-custom-file-input.min.js
  [200] http://localhost:5773/docs/4.3/assets/js/src/application.js
  [200] http://localhost:5773/docs/4.3/assets/js/src/search.js
  [200] http://localhost:5773/docs/4.3/assets/js/src/ie-emulation-modes-warning.js

ERROR: Detected 2 broken links. Scanned 47 links in 0.167 seconds.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! bootstrap@4.3.1 linkinator: `linkinator _gh_pages/ --skip "^(?!http://localhost)"`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the bootstrap@4.3.1 linkinator script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\xmr\AppData\Roaming\npm-cache\_logs\2019-02-18T08_58_54_326Z-debug.log

C:\Users\xmr\Desktop\bootstrap>npm run linkinator

> bootstrap@4.3.1 linkinator C:\Users\xmr\Desktop\bootstrap
> linkinator _gh_pages --recurse --skip "^(?!http://localhost)"

🏊‍♂️ crawling _gh_pages
  [200] http://localhost:5163

 Scanning http://localhost:5163
  [SKP] https://getbootstrap.com/
  [404] http://localhost:5163/docs/4.3/dist/css/bootstrap.css
  [200] http://localhost:5163/docs/4.3/assets/css/docs.min.css
  [200] http://localhost:5163/docs/4.3/assets/img/favicons/apple-touch-icon.png
  [200] http://localhost:5163/docs/4.3/assets/img/favicons/favicon-32x32.png
  [200] http://localhost:5163/docs/4.3/assets/img/favicons/favicon-16x16.png
  [200] http://localhost:5163/docs/4.3/assets/img/favicons/manifest.json
  [200] http://localhost:5163/docs/4.3/assets/img/favicons/safari-pinned-tab.svg
  [200] http://localhost:5163/favicon.ico
  [200] http://localhost:5163/

 Scanning http://localhost:5163/
  [200] http://localhost:5163/docs/4.3/getting-started/introduction/
...
```

**Comment by hensmith** [comment]
Thought I'd take the time to give this one a little boost!

Today I evaluated moving my employer's site to this from broken-link-checker, since that package seems to be a little abandoned. This issue ended up being what blocked the migration. For our use-case, when this check fails it's a very time-sensitive problem. This makes it super important to be able to immediately see what links are broken, what pages failed, and the relationship between the two.

I'm only posting this in the spirit of trying to contribute positively and share some info, though, and intend absolutely no entitled "DO FREE WORK FOR MY EMPLOYER'S USE-CASE AT ONCE" kind of sentiment by it! 😇😆

**Comment by JustinBeckwith** [comment]
@hensmith thanks for letting me know this is a thing that would help!  I am however, a little confused.  As far as I understand the issue right now, I just do a depth first search instead of a breadth first search.  So instead of doing all links on one page, then queueing a list to check next - we go all the way down a rabbit hole, and come back up.  I think this issue, as currently stated, is largely focused on the order - not the data returned.

> This makes it super important to be able to immediately see what links are broken, what pages failed, and the relationship between the two.

You should be able to see all that!  Do you have pages that aren't getting crawled?

**Comment by JustinBeckwith** [comment]
So - I'm 90% sure I fixed this in #35 😆 Instead of doing a serial scan with a depth first strategy, we're now parallelizing requests, which lets me add tasks in a more breadth first style.  Link results are now output as they scanned, and the final results are grouped under the parent scanned page.  Please do take a look, and let me know if this looks right.  Happy to tweak things as we go :)
